baseapp/baseapp.go

```diff
@@ -733,38 +740,93 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 	// and we're in DeliverTx. Note, runMsgs will never return a reference to a
 	// Result if any single message fails or does not have a registered Handler.
 	result, err = app.runMsgs(runMsgCtx, msgs, mode)
-	if err == nil {
-		// Run optional postHandlers.
-		//
-		// Note: If the postHandler fails, we also revert the runMsgs state.
+
+	if err != nil {
+		// Case 1: the msg errors and the post handler is not set.
+		if app.postHandler == nil {
+			return gInfo, result, anteEvents, priority, err
+		}
+
+		// Case 2: tx errors and the post handler is set. Run PostHandler and revert state from runMsgs
 		if app.postHandler != nil {
+			// Run optional postHandlers with a context branched off the ante handler ctx
+			postCtx, postCache := app.cacheTxContext(ctx, txBytes)
+
 			// The runMsgCtx context currently contains events emitted by the ante handler.
 			// We clear this to correctly order events without duplicates.
 			// Note that the state is still preserved.
-			postCtx := runMsgCtx.WithEventManager(sdk.NewEventManager())
+			postCtx = runMsgCtx.WithEventManager(sdk.NewEventManager())
 
 			newCtx, err := app.postHandler(postCtx, tx, mode == runTxModeSimulate, err == nil)
 			if err != nil {
-				return gInfo, nil, anteEvents, priority, err
+				// return result in case the pointer has been modified by the post decorators
+				return gInfo, result, events, priority, err
 			}
 
-			result.Events = append(result.Events, newCtx.EventManager().ABCIEvents()...)
-		}
+			if mode == runTxModeDeliver {
+				// When block gas exceeds, it'll panic and won't commit the cached store.
+				consumeBlockGas()
 
-		if mode == runTxModeDeliver {
-			// When block gas exceeds, it'll panic and won't commit the cached store.
-			consumeBlockGas()
+				postCache.Write()
+			}
 
-			msCache.Write()
+			if len(anteEvents) > 0 && (mode == runTxModeDeliver || mode == runTxModeSimulate) {
+				// append the events in the order of occurrence
+				postEvents = newCtx.EventManager().ABCIEvents()
+				events = make([]abci.Event, len(anteEvents)+len(postEvents))
+				copy(events[:len(anteEvents)], anteEvents)
+				copy(events[len(anteEvents):], postEvents)
+				result.Events = append(result.Events, events...)
+			} else {
+				events = make([]abci.Event, len(postEvents))
+				copy(events, postEvents)
+				result.Events = append(result.Events, postEvents...)
+			}
```

Comment:
This is really messy IMO. `runTx` is arguably the most important method of `BaseApp` and the state machine in general and it's already not written in the most friendly and intuitive way. These changes exacerbate that IMO.

Also the change from `anteEvents` to `events` is invalid. It should remain `anteEvents` because that's what we're returning.